### PR TITLE
TST make sklearn integration test compatible with 0.24

### PR DESCRIPTION
**Pull request by glemaitre**
Modify the integration test for scikit-learn. Some deprecations have been raised in 0.23 and the upcoming release of 0.24 will break this test.

### CODE REVIEW

File: tests/python_package_test/test_sklearn.py [hunk 1]
```diff
@@ -458,22 +458,10 @@ def test_sklearn_integration(self):
         # we cannot use `check_estimator` directly since there is no skip test mechanism
         for name, estimator in ((lgb.sklearn.LGBMClassifier.__name__, lgb.sklearn.LGBMClassifier),
                                 (lgb.sklearn.LGBMRegressor.__name__, lgb.sklearn.LGBMRegressor)):
-            check_parameters_default_constructible(name, estimator)
             # we cannot leave default params (see https://github.com/microsoft/LightGBM/issues/833)
             estimator = estimator(min_child_samples=1, min_data_in_bin=1)
-            for check in _yield_all_checks(name, estimator):
-                check_name = check.func.__name__ if hasattr(check, 'func') else check.__name__
-                if check_name == 'check_estimators_nan_inf':
-                    continue  # skip test because LightGBM deals with nan
```
Comment (glemaitre) [inline tests/python_package_test/test_sklearn.py:467]: This should be handled by the tag `allow_nan`.

File: tests/python_package_test/test_sklearn.py [hunk 2]
```diff
@@ -458,22 +458,10 @@ def test_sklearn_integration(self):
...
-                elif check_name == "check_no_attributes_set_in_init":
```
Comment (glemaitre) [inline tests/python_package_test/test_sklearn.py:468]: This should be handle by the `xfail` that I added

File: tests/python_package_test/test_sklearn.py [hunk 3]
```diff
@@ -458,22 +458,10 @@ def test_sklearn_integration(self):
...
-                    # skip test because scikit-learn incorrectly asserts that
-                    # private attributes cannot be set in __init__
-                    # (see https://github.com/microsoft/LightGBM/issues/2628)
-                    continue
-                try:
-                    check(name, estimator)
-                except SkipTest as message:
-                    warnings.warn(message, SkipTestWarning)
+            check_parameters_default_constructible(name, estimator)
```
Comment (glemaitre) [inline tests/python_package_test/test_sklearn.py:463]: The parameter `estimator` needs to be an instance and not a class as announce in the deprecation in 0.23

File: tests/python_package_test/test_sklearn.py [hunk 4]
```diff
@@ -455,25 +455,12 @@ def test_feature_importances_type(self):
     # sklearn <0.19 cannot accept instance, but many tests could be passed only with min_data=1 and min_data_in_bin=1
     @unittest.skipIf(sk_version < '0.19.0', 'scikit-learn version is less than 0.19')
```
Comment (glemaitre) [inline tests/python_package_test/test_sklearn.py:456]: I am not sure yet but we might need `0.23` to use the latest testing mechanism.

File: tests/python_package_test/test_sklearn.py [hunk 5]
```diff
@@ -1149,3 +1127,24 @@ def test_check_is_fitted(self):
...
+
+def _tested_estimators():
+    for Estimator in [lgb.sklearn.LGBMClassifier, lgb.sklearn.LGBMRegressor]:
+        yield Estimator()
+
+
+@pytest.mark.skipif(
+    sk_version < "0.23.0", reason="scikit-learn version is less than 0.23"
+)
+@parametrize_with_checks(list(_tested_estimators()))
```
Comment (glemaitre) [inline tests/python_package_test/test_sklearn.py:1140]: Only the `parametrize_with_checks` was using `_xfail` in 0.23. It has the advantage to show all the test from `_yield` as with `parametrize` of `pytest`.

Comment (glemaitre) [reply tests/python_package_test/test_sklearn.py:1140]: However, it needs to be a pure function not embedded within a class

File: tests/python_package_test/test_sklearn.py [hunk 6]
```diff
@@ -4,24 +4,37 @@
...
-from sklearn.exceptions import SkipTestWarning
 from sklearn.metrics import log_loss, mean_squared_error
 from sklearn.model_selection import GridSearchCV, RandomizedSearchCV, train_test_split
 from sklearn.multioutput import (MultiOutputClassifier, ClassifierChain, MultiOutputRegressor,
                                  RegressorChain)
-from sklearn.utils.estimator_checks import (_yield_all_checks, SkipTest,
-                                            check_parameters_default_constructible)
 from sklearn.utils.validation import check_is_fitted
 
+from sklearn.utils.estimator_checks import check_parameters_default_constructible
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:20]: Please move this line before `from sklearn.utils.validation ...` to keep alphabetical order for imports.

File: tests/python_package_test/test_sklearn.py [hunk 7]
```diff
@@ -4,24 +4,37 @@
...
+
+try:
+    from pkg_resources import parse_version  # type: ignore
+except ImportError:
+    # setuptools not installed
+    parse_version = LooseVersion  # type: ignore
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:26]: Seems we should import `LooseVersion` first, shouldn't we?

File: tests/python_package_test/test_sklearn.py [hunk 8]
```diff
@@ -1149,3 +1139,49 @@ def test_check_is_fitted(self):
...
+        sk_version < parse_version("0.21"), reason="scikit-learn version is less than 0.21"
+    )
+    @pytest.mark.parametrize(
+        "estimator, check",
+        _generate_checks_per_estimator(_yield_all_checks, _tested_estimators()),
+    )
+    def test_sklearn_integration(estimator, check):
+        xfail_checks = estimator._get_tags()["_xfail_checks"]
+        check_name = check.__name__ if hasattr(check, "__name__") else check.func.__name__
+        if xfail_checks and check_name in xfail_checks:
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:1166]: With `get()` it looks a bit safer.
```suggestion
        xfail_checks = estimator._get_tags().get("_xfail_checks", {})
        check_name = check.__name__ if hasattr(check, "__name__") else check.func.__name__
        if check_name in xfail_checks:
```

File: tests/python_package_test/test_sklearn.py [hunk 9]
```diff
@@ -1149,3 +1139,49 @@ def test_check_is_fitted(self):
...
+        check(name, estimator)
+else:
+    @parametrize_with_checks(list(_tested_estimators()))
+    def test_sklearn_integration(estimator, check, request):
+        estimator.set_params(min_child_samples=1, min_data_in_bin=1)
+        check(estimator)
+
+
+@pytest.mark.skipif(
+    not (sk_version < parse_version("0.24")),
```
Comment (StrikerRUS) [inline tests/python_package_test/test_sklearn.py:1180]: I believe this is more readable.
```suggestion
    (sk_version >= parse_version("0.24")),
```

**Review by StrikerRUS — COMMENTED** [review]
@glemaitre Thank you very much for updating LightGBM to follow the most recent scikit-learn API changes! Really appreciate it!

Especially thanks a lot for preserving tests for old versions, because we try to target the most wide range of versions.

Generally LGTM! Just please consider checking a few my minor comments below.

File: tests/python_package_test/test_sklearn.py (hunk 8 above)
Comment (glemaitre) [reply tests/python_package_test/test_sklearn.py:1166]: `"_xfail_checks"` is always defined because it is a default tag. At least, if there is a breaking change in the future, one could detect it with the failure. WDYT?

File: tests/python_package_test/test_sklearn.py (hunk 7 above)
Comment (glemaitre) [reply tests/python_package_test/test_sklearn.py:26]: Actually, you are importing setuptools in `setup.py` so we can safely import from `pkg_resources`.

File: tests/python_package_test/test_sklearn.py (hunk 8 above)
Comment (StrikerRUS) [reply tests/python_package_test/test_sklearn.py:1159]: Ah, so `_xfail_checks` is defined (with empty dict?) in BaseEstimator, if I got it correctly? Then it makes sense!

Then another question. Why do we need `if xfail_checks and ...` guard?

**Comment by StrikerRUS** [comment]
Fixes #2947.

**Review by StrikerRUS — APPROVED** [review]
Manny thanks @glemaitre !
